pinc/ProofProject.inc

```diff
@@ -0,0 +1,362 @@
+        $sql = "
+            SELECT state, $user_column
+            FROM $this->projectid
+            WHERE image = '$page_name'
+        ";
+        $res = DPDatabase::query($sql);
+        $row = mysqli_fetch_assoc($res);
+
+        $current_page_state = $row["state"];
+        if ($expected_page_state != $current_page_state) {
+            $err = sprintf(
+                _('Warning: Page  "%1$s" is no longer in state "%2$s"; it is now in state "%3$s".'),
+                $page_name,
+                $expected_page_state,
+                $current_page_state
+            );
+            throw new ProjectPageInconsistentStateException($err);
+        }
+
+        if ($row[$user_column] != $pguser) {
+            throw new PageNotOwnedException(_("You do not own this page."));
+        }
+
+        $this->page_name = $page_name;
+        $this->page_state = $current_page_state;
+    }
+
+    public function return_to_round(): void
+    {
+        global $pguser;
+
+        $this->validate_page(false, "returnToRound");
+        $setters = join(", ", [
+            set_col_str("state", $this->round->page_avail_state),
+            set_col_num($this->round->time_column_name, time()),
+        ]);
+        _Page_UPDATE($this->projectid, $this->page_name, $setters);
+        _log_page_event($this->projectid, $this->page_name, 'returnToRound', $pguser, $this->round);
+        _project_adjust_n_available_pages($this->projectid, +1);
+    }
+
+    public function attempt_save_as_done(): void
+    {
+        global $pguser;
+
+        $this->validate_page(false, "saveAsDone");
+        $text_checker = new TextChecker($this->project);
+        $this->page_text = $text_checker->get_valid_text();
+        if ($this->round->has_a_daily_page_limit()) {
+            $pre_save_dpl_count = get_dpl_count_for_user_in_round($pguser, $this->round);
+            if ($pre_save_dpl_count >= $this->round->daily_page_limit) {
+                // The user has already reached their limit of this kind of page.
+                $this->_save_as_in_progress();
+                $sentence = sprintf(
+                    _("Your page was saved as 'In Progress' rather than 'Done', because you have already reached the daily page limit for %s."),
+                    $this->round->id
+                );
+                throw new DailyLimitExceededException($sentence);
+            } else {
+                $this->save_as_done();
+                if ($pre_save_dpl_count + 1 >= $this->round->daily_page_limit) {
+                    $sentence = sprintf(
+                        _("Your page has been saved as 'Done'. However, you have now reached the daily page limit for %s."),
+                        $this->round->id
+                    );
+                    throw new DailyLimitReachedException($sentence);
+                }
+            }
+        } else {
+            $this->save_as_done();
+        }
+    }
+
+    private function save_as_done(): void
```

Comment:
is there a reason some private functions start with `_` and others don't?

Reply:
Not really. I did that with _save_as_in_progress so I could re-use the same function for the normal save_as_in_progress and save_as_in_progress_and_revert_to_original. It would perhaps be better to use a more obviously different name. Or possibly do save_as_in_progress_and_revert_to_original by first doing save_as_in_progress and then getting the original text in another api call.